src/cargo/util/graph.rs

```diff
@@ -165,8 +165,12 @@ impl<'s, N: Eq + Ord + Clone + 's, E: Default + Clone + 's> Graph<N, E> {
                 ));
             }
             let last = result.last().unwrap().0;
-            // fixme: this may sometimes be wrong when there are cycles.
-            if !fn_edge(&self, last).next().is_none() {
+            let set: Vec<_> = result.iter().map(|(k, _)| k).collect();
+            if !fn_edge(&self, last)
```

Comment:
Were these changes meant to be in the commit for adding a test?

Reply:
It seems more like correcting the test than fixing the implementation, despite which function it was living in. If you would like me to move it and I need to re-base, I will fix it.

Reply:
Ah, I missed the debug assertion part.  ick.